Package request: AssetRipper

**Project description**

A tool for extracting Unity serialized files.

**Metadata**

* homepage URL: https://assetripper.github.io/AssetRipper/
* source URL: https://github.com/AssetRipper/AssetRipper
* license: GPL 3.0
* platforms: unix

**My attempt**

```nix
{ fetchFromGitHub
, buildDotnetModule
, lib
, dotnetCorePackages
, gcc
, ncurses
, zlib
} :

let
  vmr = dotnetCorePackages.dotnet_8.vmr;
  sdk = dotnetCorePackages.dotnet_8.sdk;
in buildDotnetModule rec {
  pname = "asset-ripper";
  version = "1.0.9";
  src = fetchFromGitHub {
    owner = "UlyssesZh";
    repo = "AssetRipper";
    # Change to `rev = version;` when AssetRipper/AssetRipper#1244 gets merged and released.
    rev = "936aafe62c66618a6778b370e76c7e527289fcf0";
    hash = "sha256-RFFu93Tlq8kwo4v3WNKoElKHVR7+w802tDy9r+x+gHU=";
  };
  nativeBuildInputs = [ gcc vmr zlib ncurses ];
  dotnet-sdk = sdk;
  projectFile = "Source/AssetRipper.GUI.Free/AssetRipper.GUI.Free.csproj"; # #220985
  selfContainedBuild = true;
  nugetDeps = ./deps.nix;
  meta = with lib; {
    description = "A tool for extracting assets from Unity serialized files and asset bundles and converting them into the native Unity engine format.";
    homepage = "https://assetripper.github.io/AssetRipper/";
    license = licenses.gpl3;
    maintainers = with maintainers; [ ulysseszhan ];
    mainProgram = "AssetRipper";
    platforms = platforms.all;
  };
}
```
Failed with compilation errors:
```text
EXEC : error : Unable to load shared library 'objwriter' or one of its dependencies. In order to help diagnose loading problem
s, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: [/build/sourc
e/Source/AssetRipper.GUI.Free/AssetRipper.GUI.Free.csproj]
  /build/tmp.hknPKx4wii/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.2/tools/objwriter.so: cannot open sh
ared object file: No such file or directory
  /build/tmp.hknPKx4wii/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.2/tools/libobjwriter.so: cannot open
 shared object file: No such file or directory
  /build/tmp.hknPKx4wii/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.2/tools/objwriter: cannot open shared object file: No such file or directory
  /build/tmp.hknPKx4wii/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.2/tools/libobjwriter: cannot open shared object file: No such file or directory
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

> have you tried with .net 6 ?

I tried .NET 6, and it has this error:
```
error NETSDK1045: The current .NET SDK does not support targeting .NET 8.0.  Either target .NET 6.0 or lower, or use a version of the .NET SDK that supports .NET 8.0.
```
According to the error, I can only use .NET 8.

> Please try to use `tag` instead of specific commit in `fetchFromGitHub`

At the time of me initially packaging it, I encountered bugs and I submitted a pr to upstream. The pr was not merged at that time, so I used my own commit as rev temporarily. Now it has already been merged and included in a new release.

Just now I tried this:
```nix
{ fetchFromGitHub
, buildDotnetModule
, lib
, dotnetCorePackages
, gcc
, ncurses
, zlib
} :

let
  vmr = dotnetCorePackages.dotnet_8.vmr;
  sdk = dotnetCorePackages.dotnet_8.sdk;
in buildDotnetModule rec {
  pname = "asset-ripper";
  version = "1.0.11";
  src = fetchFromGitHub {
    owner = "AssetRipper";
    repo = "AssetRipper";
    rev = version;
    hash = "sha256-dfc1t2xw4RBEG882Xdhh8XP4y5aLMC7jUbkeIYDenxE=";
  };
  nativeBuildInputs = [ gcc vmr zlib ncurses ];
  dotnet-sdk = sdk;
  projectFile = "Source/AssetRipper.GUI.Free/AssetRipper.GUI.Free.csproj"; # #220985
  selfContainedBuild = true;
  nugetDeps = ./deps.nix;
  meta = with lib; {
    description = "A tool for extracting assets from Unity serialized files and asset bundles and converting them into the native Unity engine format.";
    homepage = "https://assetripper.github.io/AssetRipper/";
    license = licenses.gpl3;
    maintainers = with maintainers; [ ulysseszhan ];
    mainProgram = "AssetRipper";
    platforms = platforms.all;
  };
}
```
The error is now
```
asset-ripper> EXEC : error : Unable to load shared library 'jitinterface_x64' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: [/build/source/Source/AssetRipper.GUI.Free/AssetRipper.GUI.Free.csproj]
asset-ripper>   jitinterface_x64.so: cannot open shared object file: No such file or directory
asset-ripper>   libjitinterface_x64.so: cannot open shared object file: No such file or directory
asset-ripper>   jitinterface_x64: cannot open shared object file: No such file or directory
asset-ripper>   libjitinterface_x64: cannot open shared object file: No such file or directory
asset-ripper>   System.DllNotFoundException: Unable to load shared library 'jitinterface_x64' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable:
asset-ripper>   jitinterface_x64.so: cannot open shared object file: No such file or directory
asset-ripper>   libjitinterface_x64.so: cannot open shared object file: No such file or directory
asset-ripper>   jitinterface_x64: cannot open shared object file: No such file or directory
asset-ripper>   libjitinterface_x64: cannot open shared object file: No such file or directory
asset-ripper>      at System.Runtime.InteropServices.NativeLibrary.LoadLibErrorTracker.Throw(String) + 0x46
asset-ripper>      at System.Runtime.InteropServices.NativeLibrary.Load(String, Assembly, Nullable`1) + 0xd7
asset-ripper>      at Internal.JitInterface.JitConfigProvider.<>c__DisplayClass5_0.<Initialize>b__0(String libName, Assembly assembly, Nullable`1 searchPath) + 0x157
asset-ripper>      at System.Runtime.InteropServices.NativeLibrary.LoadLibraryCallbackStub(String, Assembly, Boolean, UInt32) + 0xa0
asset-ripper>      at Internal.Runtime.CompilerHelpers.InteropHelpers.FixupModuleCell(InteropHelpers.ModuleFixupCell*) + 0x71
asset-ripper>      at Internal.Runtime.CompilerHelpers.InteropHelpers.ResolvePInvokeSlow(InteropHelpers.MethodFixupCell*) + 0x35
asset-ripper>      at Internal.JitInterface.CorInfoImpl.GetJitHost(IntPtr) + 0x30
asset-ripper>      at Internal.JitInterface.CorInfoImpl.Startup(CORINFO_OS) + 0x1f
asset-ripper>      at ILCompiler.RyuJitCompilationBuilder.ToCompilation() + 0x413
asset-ripper>      at ILCompiler.Program.Run() + 0x2825
asset-ripper>      at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass227_0.<.ctor>b__0(ParseResult result) + 0x310
asset-ripper> /build/tmp.gP47nywV9f/.nuget/packages/microsoft.dotnet.ilcompiler/8.0.2/build/Microsoft.NETCore.Native.targets(308,5): error MSB3073: The command ""/build/tmp.gP47nywV9f/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.2/tools/ilc" @"../0Bins/obj/AssetRipper.GUI.Free/Release/linux-x64/native/AssetRipper.GUI.Free.ilc.rsp"" exited with code 1. [/build/source/Source/AssetRipper.GUI.Free/AssetRipper.GUI.Free.csproj]
```
Seems related to #280923.

--- a/pkgs/by-name/as/assetripper/package.nix
+++ b/pkgs/by-name/as/assetripper/package.nix
@@ -1,0 +1,87 @@
+{
+  lib,
+  stdenv,
+  autoPatchelfHook,
+  buildDotnetModule,
+  fetchFromGitHub,
+  dbus,
+  dotnetCorePackages,
+}:
+
+buildDotnetModule (finalAttrs: {
+  pname = "assetripper";
+  version = "1.3.0";
+
+  src = fetchFromGitHub {
+    owner = "AssetRipper";
+    repo = "AssetRipper";
+    tag = finalAttrs.version;
+    hash = "sha256-ixXWbygFhvOjld+YRLIhkO3cgDNkQsbivri2pjU4rgM=";
+  };
+
+  postPatch = ''
+    sed 's@Path.Join(ExecutingDirectory, "temp",@Path.Join(Path.GetTempPath(), "AssetRipper",@' \
+      -i Source/AssetRipper.IO.Files/Utils/TemporaryFileStorage.cs
+  '';
+
+  buildInputs = [
+    dbus
+    (lib.getLib stdenv.cc.cc)
+  ];
+
+  nativeBuildInputs = [ autoPatchelfHook ];
+
+  # Prevent automatic patching of all files. This is necessary as applying
+  # autoPatchelf indiscriminately causes dangling references to openssl and
+  # icu4c in AssetRipper.GUI.Free
+  dontAutoPatchelf = true;
+
+  # Avoid IOException on startup
+  makeWrapperArgs = [
+    "--add-flags"
+    "--log=false"
+  ];
+
+  # Make the main executable available under a more intuitive name.
+  postInstall = ''
+    mkdir -p $out/bin
+    ln -rs $out/bin/AssetRipper.GUI.Free $out/bin/AssetRipper
+  '';
+
+  # Patch some prebuilt libraries fetched via NuGet.
+  fixupPhase = ''
+    runHook preFixup
+
+    autoPatchelf $out/lib/${finalAttrs.pname}/libnfd.so
+    autoPatchelf $out/lib/${finalAttrs.pname}/libTexture2DDecoderNative.so
+
+    runHook postFixup
+  '';
+
+  projectFile = "Source/AssetRipper.GUI.Free/AssetRipper.GUI.Free.csproj";
+
+  # Error: "PublishTrimmed is implied by native compilation and cannot be disabled."
+  # We need to override the project settings and disable native AoT compilation
+  # as this is incompatible with PublishTrimmed.
+  dotnetInstallFlags = [ "-p:PublishAot=false" ];
+
+  nugetDeps = ./deps.json;
+
+  executables = [ "AssetRipper.GUI.Free" ];
+
+  dotnet-sdk = dotnetCorePackages.sdk_9_0;
+  dotnet-runtime = finalAttrs.dotnet-sdk.aspnetcore;
+
+  meta = {
+    description = "Tool for extracting assets from Unity serialized files and asset bundles";
+    homepage = "https://github.com/AssetRipper/AssetRipper";
+    license = lib.licenses.gpl3Only;
+    mainProgram = "AssetRipper";
+    maintainers = with lib.maintainers; [ YoshiRulz ];
+    platforms = lib.platforms.unix;
+    sourceProvenance = with lib.sourceTypes; [
+      fromSource
+      binaryNativeCode # libraries fetched by NuGet
+    ];
+  };
+})

